### PR TITLE
[vmvx] Add round to VM dialect and fixup bytecode_op_table generation

### LINKED ISSUE #7618: Missing legalization for `mhlo.round_nearest_afz`
Repro:
```
from iree.compiler import compile_str

CODE = """
module  {
  func @main(%arg0: tensor<1x4xf32>) -> tensor<1x4xf32> {
    %0 = "mhlo.round_nearest_afz"(%arg0) : (tensor<1x4xf32>) -> tensor<1x4xf32>
    return %0 : tensor<1x4xf32>
  }
}"""

compiled_flatbuffer = compile_str(CODE, target_backends=["dylib"], input_type="mhlo")
```

Get:
```
$ python t.py
Traceback (most recent call last):
  File "t.py", line 11, in <module>
    compiled_flatbuffer = compile_str(CODE, target_backends=["dylib"], input_type="mhlo")
  File "/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/core.py", line 262, in compile_str
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/binaries.py", line 201, in invoke_immediate
    raise CompilerToolError(process)
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool ireec
Diagnostics:
<stdin>:4:10: error: failed to legalize operation 'mhlo.round_nearest_afz' that was explicitly marked illegal
    %0 = "mhlo.round_nearest_afz"(%arg0) : (tensor<1x4xf32>) -> tensor<1x4xf32>
         ^
<stdin>:3:3: note: called from
  func @main(%arg0: tensor<1x4xf32>) -> tensor<1x4xf32> {
  ^
<stdin>:2:1: error: conversion from source -> vm failed
module  {
^


Invoked with:
 ireec /usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/../_mlir_libs/ireec - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=dylib --iree-mlir-to-vm-bytecode-module --iree-llvm-embedded-linker-path=/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false
```

### PR DESCRIPTION
**Pull request by rsuderman**
Round is a reasonably common operation for models, added to the VM
dialect for round operations. This includes some cleanup for generating
the bytecode table, along with missing some declarations of the ErfF32
function.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td [hunk 1]
```diff
@@ -63,20 +63,21 @@ def VM_OPC_RsqrtF32              : VM_OPC<0x26, "RsqrtF32">;
...
-def VM_OPC_CmpLTF32O             : VM_OPC<0x2E, "CmpLTF32O">;
-def VM_OPC_CmpLTF32U             : VM_OPC<0x2F, "CmpLTF32U">;
-def VM_OPC_CmpLTEF32O            : VM_OPC<0x30, "CmpLTEF32O">;
-def VM_OPC_CmpLTEF32U            : VM_OPC<0x31, "CmpLTEF32U">;
-def VM_OPC_CmpNaNF32             : VM_OPC<0x32, "CmpNaNF32">;
-
-def VM_OPC_BufferLoadF32         : VM_OPC<0x33, "BufferLoadF32">;
-def VM_OPC_BufferStoreF32        : VM_OPC<0x34, "BufferStoreF32">;
-def VM_OPC_BufferFillF32         : VM_OPC<0x35, "BufferFillF32">;
+def VM_OPC_RoundF32              : VM_OPC<0x2A, "RoundF32">;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td:66]: put round up after Floor instead (here and in all other files to keep things consistent)

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td [hunk 2]
```diff
@@ -63,20 +63,21 @@ def VM_OPC_RsqrtF32              : VM_OPC<0x26, "RsqrtF32">;
...
+
+def VM_OPC_CmpEQF32O             : VM_OPC<0x2B, "CmpEQF32O">;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td:68]: changing these breaks binary compatibility - if you assign it 0x36 instead (they don't have to be in order) we can do the reordering at some point when we want to break things and then we don't need a version bump here (adding opcodes is backwards compatible)

File: runtime/src/iree/vm/ops.h [hunk 3]
```diff
@@ -312,6 +312,8 @@ static inline float vm_sqrt_f32(float operand) { return sqrtf(operand); }
 static inline float vm_tanh_f32(float operand) { return tanhf(operand); }
 static inline float vm_erf_f32(float operand) { return erff(operand); }
 
+static inline float vm_round_f32(float operand) { return round(operand); }
```
Comment (benvanik) [inline runtime/src/iree/vm/ops.h:315]: use roundf (round is double)

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td]: Fixed.

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td (hunk 2 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/IR/VMOpcodesF32.td]: Sounds good. Was not certain whether binary-compatibility or order should be prioritized.

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td [hunk 4]
```diff
@@ -2334,6 +2334,13 @@ def VM_ErfF64Op : VM_UnaryArithmeticOp<F64, "erf.f64", VM_OPC_ErfF64, [VM_ExtF64
   let summary = [{computes the error function of the specified value}];
 }
 
+def VM_RoundF32Op : VM_UnaryArithmeticOp<F32, "round.f32", VM_OPC_RoundF32, [VM_ExtF32]> {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td:2337]: reorder to below floor

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp [hunk 5]
```diff
@@ -4642,6 +4642,8 @@ void populateVMToEmitCPatterns(ConversionTarget &conversionTarget,
                                                          "vm_tanh_f32");
   patterns.add<GenericOpConversion<IREE::VM::ErfF32Op>>(typeConverter, context,
                                                         "vm_erf_f32");
+  patterns.add<GenericOpConversion<IREE::VM::RoundF32Op>>(
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp:4645]: reorder to below floor

File: compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp [hunk 6]
```diff
@@ -119,7 +119,10 @@ void populateMathToVMPatterns(MLIRContext *context,
       UnaryArithmeticOpConversion<math::TanhOp, IREE::VM::TanhF32Op,
                                   IREE::VM::TanhF64Op>,
       UnaryArithmeticOpConversion<math::ErfOp, IREE::VM::ErfF32Op,
-                                  IREE::VM::ErfF64Op>>(typeConverter, context);
+                                  IREE::VM::ErfF64Op>,
+      UnaryArithmeticOpConversion<math::RoundOp, IREE::VM::RoundF32Op,
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp:123]: reorder to below floor

File: runtime/src/iree/vm/bytecode_dispatch.c [hunk 7]
```diff
@@ -2027,6 +2027,7 @@ static iree_status_t iree_vm_bytecode_dispatch(
       DISPATCH_OP_EXT_F32_UNARY_F32(SqrtF32, vm_sqrt_f32);
       DISPATCH_OP_EXT_F32_UNARY_F32(TanhF32, vm_tanh_f32);
       DISPATCH_OP_EXT_F32_UNARY_F32(ErfF32, vm_erf_f32);
+      DISPATCH_OP_EXT_F32_UNARY_F32(RoundF32, vm_round_f32);
```
Comment (benvanik) [inline runtime/src/iree/vm/bytecode_dispatch.c:2030]: reorder to below floor

File: runtime/src/iree/vm/ops.h [hunk 8]
```diff
@@ -312,6 +312,8 @@ static inline float vm_sqrt_f32(float operand) { return sqrtf(operand); }
 static inline float vm_tanh_f32(float operand) { return tanhf(operand); }
 static inline float vm_erf_f32(float operand) { return erff(operand); }
 
+static inline float vm_round_f32(float operand) { return roundf(operand); }
```
Comment (benvanik) [inline runtime/src/iree/vm/ops.h:315]: reorder to below floor

**Review by benvanik — DISMISSED** [review]

File: compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp (hunk 6 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp:129]: Floor actually wasn't included. Added in the other operations but not sure how that factors into total ordering.

File: runtime/src/iree/vm/bytecode_dispatch.c (hunk 7 above)
Comment (rsuderman) [reply runtime/src/iree/vm/bytecode_dispatch.c]: Done.

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td (hunk 4 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td]: Done.

File: compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp (hunk 5 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp]: Done.

File: compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp (hunk 6 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp:129]: Oh this list is weird? It's almost alphabetical? Maybe group the trig functions together and them order the rest alphabetically?

```
// Trigonometric functions.
patterns.insert< ...

// Other math functions.
patterns.insert< ...
```

or just add `round` in this PR, then in a follow-up add the other conversions that were missing

Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/VM/Conversion/MathToVM/ConvertMathToVM.cpp]: Reordered to the same as op declarations.

File: tests/e2e/xla_ops/BUILD [hunk 9]
```diff
@@ -312,7 +313,6 @@ iree_check_single_backend_test_suite(
         include = ["*.mlir"],
         exclude = [
             "exponential_fp16.mlir",
-            "round.mlir",  # TODO(#7618): Fix and enable (math.copysign -> VM)
```
Comment (ScottTodd) [inline tests/e2e/xla_ops/BUILD:315]: tagging https://github.com/iree-org/iree/issues/7618 here (maybe all fixed now?)

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**